synapseclient/client.py

```diff
@@ -310,9 +312,40 @@ def __init__(
         else:
             self._requests_session_async_synapse = {}
 
+        span_dict: Dict[httpx.Request, trace.Span] = {}
+
+        def log_request(request: httpx.Request) -> None:
+            """
+            Log the HTTPX request to an otel span.
+
+            Arguments:
+                request: The HTTPX request object.
+            """
+            # Don't log the query string as it will contain tokens
+            url_without_query_string: httpx.URL = request.url.copy_with(query=None)
+            span = tracer.start_span(
+                f"{request.method} {url_without_query_string}", kind=SpanKind.CLIENT
+            )
+            span.set_attributes(
+                {"url": str(url_without_query_string), "http.method": request.method}
+            )
+            span_dict.update({request: span})
+
+        def log_response(response: httpx.Response) -> None:
+            """
+            Log the HTTPX response to an otel span.
+
+            Arguments:
+                response: The HTTPX response object.
+            """
+            span = span_dict.pop(response.request)
+            span.set_attribute("http.response.status_code", response.status_code)
+            span.end()
+
+        event_hooks = {"request": [log_request], "response": [log_response]}
         httpx_timeout = httpx.Timeout(70)
         self._requests_session_storage = requests_session_storage or httpx.Client(
-            timeout=httpx_timeout
+            timeout=httpx_timeout, event_hooks=event_hooks
```

Comment:
Well, this is convenient.  Are we no longer using `requests`?

Reply:
> Are we no longer using requests?

In all new code we're going to be using the httpx client. I cannot replace the requests library yet (For the previous code) because it would be a breaking change if someone were passing in their own client into the `Synapse()` class initializer.

> Basically, how I'm viewing this is that tracers are only provided by the httpx calls correct?

That's what I am moving towards - Yes. The caveat however is that I want a span for the overall Synapse entity doing actions. So a span is started for a Project, Folder, File, Table, Team ect.... that is being saved. Then inside of that span all network calls will have their span (So we can see start/end times) and I/O work like MD5 calculation will have a span.

![image](https://github.com/Sage-Bionetworks/synapsePythonClient/assets/17128019/3f300175-748b-4372-a459-5c3020de2337)